third_party/intel/lib/TritonGENToLLVM/Attributes.cpp

```diff
@@ -0,0 +1,436 @@
+    if (auto ArrAttr = dyn_cast<ArrayAttr>(NewAttr)) {
+      assert(ArrAttr.size() == 2 && isa<StringAttr>(ArrAttr[0]));
+      NamedAttribute NamedAttr(cast<StringAttr>(ArrAttr[0]), ArrAttr[1]);
+      addToPassThroughAttr(PassThroughAttr, NamedAttr, Ctx);
+    } else if (auto StrAttr = dyn_cast<StringAttr>(NewAttr)) {
+      NamedAttribute NamedAttr(StrAttr, UnitAttr::get(&Ctx));
+      addToPassThroughAttr(PassThroughAttr, NamedAttr, Ctx);
+    } else
+      llvm_unreachable("Unexpected attribute kind");
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// AttributeList Method Implementations
+//===----------------------------------------------------------------------===//
+
+AttributeList::AttributeList(const mlir::NamedAttrList &FnAttrs,
+                             const mlir::NamedAttrList &RetAttrs,
+                             llvm::ArrayRef<mlir::NamedAttrList> ParamAttrs)
+    : FnAttrs(FnAttrs), RetAttrs(RetAttrs), ParamAttrs(ParamAttrs) {}
+
+AttributeList &
+AttributeList::addAttrs(const AttrBuilder &FnAttrB, const AttrBuilder &RetAttrB,
+                        llvm::ArrayRef<mlir::NamedAttrList> Attrs) {
+  return addFnAttrs(FnAttrB).addRetAttrs(RetAttrB).addParamAttributes(Attrs);
+}
+
+AttributeList &AttributeList::addFnAttrs(const AttrBuilder &B) {
+  return addFnAttrs(B.getAttributes(), B.getContext());
+}
+
+AttributeList &AttributeList::addFnAttrs(const NamedAttrList &Attrs,
+                                         MLIRContext &Ctx) {
+  for (const NamedAttribute &NewFnAttr : Attrs) {
+    std::optional<NamedAttribute> ExistingFnAttr =
+        FnAttrs.getNamed(NewFnAttr.getName());
+    if (!ExistingFnAttr) {
+      FnAttrs.append(NewFnAttr);
+      continue;
+    }
+
+    // Merge the 'passthrough' attribute lists.
+    if (ExistingFnAttr->getName() == PassThroughAttrName) {
+      auto Attrs = cast<ArrayAttr>(NewFnAttr.getValue());
+      addToPassThroughAttr(*ExistingFnAttr, Attrs, Ctx);
+      FnAttrs.set(ExistingFnAttr->getName(), ExistingFnAttr->getValue());
+      continue;
+    }
+
+    llvm_unreachable("Function attribute already exists");
+  }
+
+  return *this;
+}
+
+AttributeList &AttributeList::addRetAttrs(const AttrBuilder &B) {
+  return addRetAttrs(B.getAttributes(), B.getContext());
+}
+
+AttributeList &AttributeList::addRetAttrs(const mlir::NamedAttrList &Attrs,
+                                          mlir::MLIRContext &Ctx) {
+  for (const NamedAttribute &NewNamedAttr : Attrs) {
+    std::optional<NamedAttribute> ExistingAttr =
+        RetAttrs.getNamed(NewNamedAttr.getName());
+    if (!ExistingAttr) {
+      RetAttrs.append(NewNamedAttr);
+      continue;
+    }
+    llvm_unreachable("Return value attribute already exits");
+  }
+
+  return *this;
+}
+
+AttributeList &
+AttributeList::addParamAttributes(llvm::ArrayRef<mlir::NamedAttrList> Attrs) {
+  ParamAttrs.reserve(Attrs.size());
+  llvm::append_range(ParamAttrs, Attrs);
```

Comment:
Simpler, `SmallVector` should know how to do this efficiently.
```suggestion
  ParamAttrs.append(Attrs.begin(), Attrs.end());
```

Reply:
done